src/Tools/dotnet-gcdump/DotNetHeapDump/EventPipeDotNetHeapDumper.cs

```diff
@@ -0,0 +1,238 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Graphs;
+using Microsoft.Diagnostics.Tools.GCDump.DotNetHeapDump;
+using Microsoft.Diagnostics.Tools.RuntimeClient;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Parsers.Clr;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.GCDump
+{
+    public static class EventPipeDotNetHeapDumper
+    {
+        /// <summary>
+        /// Given a factory for creating an EventPipe session with the appropriate provider and keywords turned on,
+        /// generate a GCHeapDump using the resulting events.  The correct keywords and provider name
+        /// are given as input to the Func eventPipeEventSourceFactory.
+        /// </summary>
+        /// <param name="processID"></param>
+        /// <param name="eventPipeEventSourceFactory">A delegate for creating and stopping EventPipe sessions</param>
+        /// <param name="memoryGraph"></param>
+        /// <param name="log"></param>
+        /// <param name="dotNetInfo"></param>
+        /// <returns></returns>
+        public static bool DumpFromEventPipe(int processID, MemoryGraph memoryGraph, TextWriter log, DotNetHeapInfo dotNetInfo = null)
+        {
+            var sw = Stopwatch.StartNew();
+            var dumper = new DotNetHeapDumpGraphReader(log)
+            {
+                DotNetHeapInfo = dotNetInfo
+            };
+            bool dumpComplete = false;
+            bool listening = false;
+
+            EventPipeSession gcDumpSession = null;
+            Task readerTask = null;
+            try
+            {
+                bool etwDataPresent = false;
+                TimeSpan lastEtwUpdate = sw.Elapsed;
+                EventPipeSession typeFlushSession = null;
+                var fDone = false;
+                var otherListening = false;
+                log.WriteLine("{0,5:n1}s: Creating type table flushing task", sw.Elapsed.TotalSeconds);
+                var typeTableFlushTask = Task.Factory.StartNew(() =>
+                {
+                    typeFlushSession = new EventPipeSession(processID, new List<Provider> { new Provider("Microsoft-DotNETCore-SampleProfiler") }, false);
+                    otherListening = true;
+                    log.WriteLine("{0,5:n1}s: Flushing the type table", sw.Elapsed.TotalSeconds);
```

Comment:
All instance properties/methods in the BCL are by default not thread-safe. You could put locks around your shared state or keep the code doing logging to a single thread at a time.

Reply:
The `TextWriter` I'm using here is `Console.Out` (or `TextWriter.Null` which doesn't matter) which has `TextWriter.Synchronize()` called on it when it is initialized, so it is thread-safe already.

Reply:
Sorry, I should have clarified... Stopwatch.Elapsed is the API I was referencing there : )

Reply:
Opted to switch to using a static method on `Stopwatch` to get timestamps which should alleviate this.